src/transformers/modeling_utils.py

```diff
@@ -539,15 +539,24 @@ def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
             )
 
 
-def set_initialized_submodules(model, state_dict_keys):
+def set_initialized_submodules(model, state_dict_keys, loaded=True):
     """
     Sets the `_is_hf_initialized` flag in all submodules of a given model when all its weights are in the loaded state
     dict.
     """
     for module_name, module in model.named_modules():
-        loaded_keys = [k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")]
-        if len(set(module.state_dict().keys()) - set(loaded_keys)) == 0:
-            module._is_hf_initialized = True
+        if loaded:
+            loaded_keys = [
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if len(set(module.state_dict().keys()) - set(loaded_keys)) == 0:
+                module._is_hf_initialized = loaded
+        else:
+            not_loaded_keys = [
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if len(set(module.state_dict().keys()).intersection(not_loaded_keys)) > 0:
+                module._is_hf_initialized = False
```

Comment:
We want to be able to set some module(s) as `not initialized` in the case we have some checkpoint/model weight(s) having different shape(s) - that **we can only find later**.

In this case, currently on `main`, we have:

- this block marks them as initialized (but not really initialize the weight)
- later, when we want to set the model weight by the checkpoint weight, it is skipped as the shape doesn't match (if `ignore_mismatched_sizes=True`)
  - so that/those weights are not initialized by the model's `_init_weights` and could get crazy values like `1e37`, and cause training issue.